scheds/rust/scx_layered/src/layer_core_growth.rs

```diff
@@ -243,28 +243,23 @@ impl<'a> LayerCoreOrderGenerator<'a> {
 
     fn grow_node_spread(&self) -> Vec<usize> {
         let mut cores: Vec<usize> = Vec::new();
-        let core_vecs: Vec<Vec<usize>> = self
-            .topo
-            .nodes
-            .clone()
-            .into_iter()
-            .map(|(_, x)| {
-                x.all_cores
-                    .clone()
-                    .into_iter()
-                    .flat_map(|(_, x)| {
-                        if self.topo.all_cores.len() * 2 == self.topo.all_cpus.len() {
-                            vec![x.id, x.id + self.topo.all_cores.len()]
-                        } else {
-                            vec![x.id]
-                        }
-                    })
-                    .collect()
-            })
-            .collect();
+        let mut cpus_vec: Vec<Vec<usize>> = Vec::new();
+
+        // XXX this probably does not work with nosmt.
+        // but layered doesn't work with that at all atm.
+
+        for (_, node) in self.topo.nodes.clone() {
+            let mut node_vec = Vec::new();
+            for (_, core) in node.all_cores.clone() {
+                for (cpu_id, _) in core.cpus.clone() {
+                    node_vec.push(cpu_id)
+                }
+            }
+            cpus_vec.push(node_vec);
+        }
 
-        for i in 0..core_vecs[0].len() {
-            for sub_vec in core_vecs.clone() {
+        for i in 0..cpus_vec[0].len() {
+            for sub_vec in cpus_vec.clone() {
                 if i < sub_vec.len() {
                     cores.push(sub_vec[i]);
                 }
```

Comment:
Hmm... this looks complicated and still fragile (what if not all nodes have the same number of CPUs?). Wouldn't something like the following work?
```
    let mut cores: Vec<usize> = vec![];
    let mut iters:Vec<_> = self.topo.nodes.iter().map(|x| x.iter()).collect();
    loop {
        let mut had_some = false;
        for it in &mut iters {
            if let Some(v) = it.next() {
                cores.push(*v);
                had_some = true;
            }
        }
        if !had_some {
            break;
        }
    }
```

Reply:
Oh, the above assumes that an iterator keeps returning `None` after returning `None` once which isn't guarnateed. Maybe just `.remove()` the iterator from the vec and replace `loop - if { break; }` with `while (iters.len())`?

Reply:
I did this in a different but (I think) not fragile way.